# Anna R Rogers
# <font color='red'>Introduction to Pandas Exercise</font>

The last two weeks focused on the Pandas library, specifically input of files to dataframes, indexing a selecting subsets of data, merging and joining, and output of dataframes to various files. This exercise is designed to review the materials coverd in the Week 4 and 5 lectures.

1. Import Pandas, Numpy and OS libraries

In [1]:
import pandas as pd
import numpy as np
import os

2. Read in the peanut_lines CSV file as **peanut_lines** (Use encoding='ISO-8859-1')

In [2]:
peanut_lines = pd.read_csv('peanut_lines.csv', encoding = 'ISO-8859-1')

3. Check the head of the files

In [3]:
peanut_lines.head()

NC_Accession Identity or Parentage Pedigree    FAG
0  ACI WT09-0761         ACI WT09-0761      NaN  Check
1  ACI WT11-0351         ACI WT11-0351      NaN  ol ol
2  ACI WT12-0226         ACI WT12-0226      NaN  ol ol
3  ACI WT12-0419         ACI WT12-0419      NaN  Check
4  ACI WT12-0420         ACI WT12-0420      NaN  ol ol

4. Use the .info() method to find out how many peanut_lines are in the dataframe

    **Bonus: Print the total number of entries?

In [4]:
peanut_lines.info() #Check how many entries are in the data frame (Bonus)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 859 entries, 0 to 858
Data columns (total 4 columns):
NC_Accession             466 non-null object
Identity or Parentage    466 non-null object
Pedigree                 409 non-null object
FAG                      466 non-null object
dtypes: object(4)
memory usage: 26.9+ KB


In [5]:
peanut_lines["NC_Accession"].nunique() #Use describe to check how many unique accessions of lines there are

463

There are 463 unique accessions in the file.

5. Read in the peanut_yield text file as **peanut_yield** (Use encoding='ISO-8859-1')

In [6]:
peanut_yield = pd.read_csv('peanut_yield.txt', encoding = 'ISO-8859-1', sep = '\t')
peanut_yield.head()

Year Location Name                             Label   NC_Accession  \
0  2014      LEW  ATP  Advanced Testing Program - Yield  ACI WT09-0761   
1  2014      RMT  ATP  Advanced Testing Program - Yield  ACI WT09-0761   
2  2014      WHI  ATP  Advanced Testing Program - Yield  ACI WT09-0761   
3  2015      LEW  ATP  Advanced Testing Program - Yield  ACI WT11-0351   
4  2015      RMT  ATP  Advanced Testing Program - Yield  ACI WT11-0351   

   Plot_Yield  Yield  
0         9.0   2733  
1        10.2   3085  
2        10.2   3085  
3        15.3   4632  
4        10.7   3226

6. Check the column names of peanut_yield

In [7]:
list(peanut_yield) #List the names of the columns in the frame

['Year', 'Location', 'Name', 'Label', 'NC_Accession', 'Plot_Yield', 'Yield']

7. Using 'NC_Accession', merge (or join) the two dataframes together keeping the entirety of the peanut_yield data. Name the new dataframe **peanut_data**

In [8]:
peanut_data = pd.merge(peanut_lines, peanut_yield, how = "right", on = 'NC_Accession')
#Use a merge on NC_accession, and merge right to get all the observations from peanut_yield

8. What is the average yield in 2017?

In [9]:
peanut_data['Yield'][peanut_data['Year'] == 2017].mean()
#Subset the data so that we only have 2017 data, then compute the mean

3981.6

The mean yield in 2017 was 3981.6

9. What was the average yield of the top 10 lines tested in the peanut program?

In [10]:
peanut_data.groupby(['NC_Accession']).mean().nlargest(10, columns = 'Yield')  
#First group by accession and calculate the mean.  Then use nlargest to find the top ten average yielding varieties

Year  Plot_Yield        Yield
NC_Accession                                 
N17045        2018.0   18.100000  5483.666667
N15006        2015.0   17.350000  5250.000000
N15060        2015.0   17.250000  5209.500000
N15004        2015.0   17.150000  5197.000000
N15066        2015.0   17.150000  5190.500000
N15055        2015.0   16.950000  5129.500000
N14025        2014.0   16.866667  5095.000000
PD 14002      2015.0   16.800000  5082.500000
PD 14003      2015.0   16.350000  4959.000000
N15001        2015.0   16.400000  4954.000000

10. What are the top 10 most commonly tested lines? Hint: value_counts()

In [11]:
peanut_data['NC_Accession'].value_counts()[0:10] #Get the value counts of the top ten most planted lines

N11028        34
Bailey        26
Emery         26
Wynne         26
Sugg          26
Sullivan      26
Florida-07    20
N11020        20
Bailey II     20
Gregory       18
Name: NC_Accession, dtype: int64

11. Create a dataframe that satisfy the following requirements
<br>peanut lines = 'Bailey', 'Sullivan', 'Wynne', 'Emery', 'Bailey II', 'N14023'
<br>location = 'LEW', 'RMT' #Lewiston and Rocky Mount
<br>Dataframe = line_data

In [12]:
peanut_lines = ['Bailey','Sullivan','Wynne','Emery','Bailey II','N14023'] #List of accessions
location = ['LEW', 'RMT'] #list of locations

line_data = peanut_data[(peanut_data["NC_Accession"].isin(peanut_lines)) & peanut_data["Location"].isin(location)] #Take only obs for the 6 lines in the two locations
line_data.head() #head of the line data

NC_Accession Identity or Parentage                    Pedigree  FAG  Year  \
15       Bailey    NC 12C*2 / N96076L  BC1F1-06-01-S-03-S-05: F09  + +  2010   
16       Bailey    NC 12C*2 / N96076L  BC1F1-06-01-S-03-S-05: F09  + +  2010   
18       Bailey    NC 12C*2 / N96076L  BC1F1-06-01-S-03-S-05: F09  + +  2011   
19       Bailey    NC 12C*2 / N96076L  BC1F1-06-01-S-03-S-05: F09  + +  2011   
21       Bailey    NC 12C*2 / N96076L  BC1F1-06-01-S-03-S-05: F09  + +  2012   

   Location Name                             Label  Plot_Yield  Yield  
15      LEW  ATP  Advanced Testing Program - Yield        11.7   3529  
16      RMT  ATP  Advanced Testing Program - Yield         6.6   2004  
18      LEW  ATP  Advanced Testing Program - Yield        13.6   4114  
19      RMT  ATP  Advanced Testing Program - Yield        17.1   5159  
21      LEW  ATP  Advanced Testing Program - Yield        12.8   3873

12. Export the means of each line by location to an excel file called 'NCSU_release.xlsx'

In [13]:
#Use group by to do this, then pipe to mean, then to excel file which will go in current working directory
line_data.groupby(["Location"]).mean().to_excel('NCSU_release.xlsx')
